### PR TITLE
Better Map support

### PR DESCRIPTION
**Pull request by cchantep**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Typesafe CLA](https://www.typesafe.com/contribute/cla)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [ ] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Purpose

Better support for `Map[K, V]` with JSON object, when `K` is not `String`

## Background Context

Introduces `KeyReads` and `KeyWrites` typeclasses.

### CODE REVIEW

**Comment by jtjeferreira** [comment]
Hi @cchantep just an heads up that #231 adds support for scala 2.13 MapView, and this PR should take that into consideration

**Comment by marcospereira** [comment]
@cchantep this now has some conflicts.

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala [hunk 1]
```diff
@@ -72,17 +72,22 @@ class ScalaJsonCombinatorsSpec extends Specification {
       //#jspath-traverse
 
       //val name = (JsPath \ "name").read[String] and (JsPath \ "location").read[Int]
-      latPath.toString === "/location/lat"
-      namesPath.toString === "//name"
-      firstResidentPath.toString === "/residents(0)"
+      longPath.toString === "/location/long" and {
```
Comment (cchantep) [inline docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala]: Fix warnings

File: docs/manual/working/scalaGuide/main/json/code/ScalaJsonSpec.scala [hunk 2]
```diff
@@ -298,8 +298,6 @@ class ScalaJsonSpec extends Specification {
     }
 
     "allow converting JsValue using as" in {
-
```
Comment (cchantep) [inline docs/manual/working/scalaGuide/main/json/code/ScalaJsonSpec.scala]: Fix warnings

File: play-json/jvm/src/test/scala/play/api/libs/json/JsonValidSpec.scala [hunk 3]
```diff
@@ -5,7 +5,7 @@
 package play.api.libs.json
 
 import org.specs2.mutable._
-import play.api.libs.json._
```
Comment (cchantep) [inline play-json/jvm/src/test/scala/play/api/libs/json/JsonValidSpec.scala]: Unused

File: play-json/shared/src/main/scala/play/api/libs/json/JsConstraints.scala [hunk 4]
```diff
@@ -204,10 +204,16 @@ trait ConstraintWrites {
   def pure[A](fixed: => A)(implicit wrs: Writes[A]): Writes[JsValue] =
     Writes[JsValue] { js => wrs.writes(fixed) }
 
+  @com.github.ghik.silencer.silent
+  @deprecated("Use `pruned` without `Writes[A]`", "2.8.0")
```
Comment (cchantep) [inline play-json/shared/src/main/scala/play/api/libs/json/JsConstraints.scala]: Fix warning

**Comment by cchantep** [comment]
Fixed @marcospereira

File: docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala [hunk 5]
```diff
@@ -197,8 +199,8 @@ class ScalaJsonCombinatorsSpec extends Specification {
       val json: JsValue = sampleJson
 
       json.validate[Place] match {
-        case s: JsSuccess[Place] => {
-          val place: Place = s.get
+        case JsSuccess(place, _) => {
+          val _: Place = place
```
Comment (renatocaval) [inline docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala]: because `place` is already being extracted, we can remove this line.

Comment (cchantep) [reply docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala]: Removing would raise unused warning, whereas it's interesting to have such test as example how to pattern-match on `JsResult`

Comment (renatocaval) [reply docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala]: yes, correct. Thanks for the explanation.

**Comment by cchantep** [comment]
> Looking good.
> 
> Left a minor comment.
> 
> What about @jtjeferreira comment on use MapView?

Already rebased

**Pull request merged**